src/main/java/com/atlassian/bitbucket/jenkins/internal/listener/BitbucketBuildListener.java

```diff
@@ -0,0 +1,38 @@
+package com.atlassian.bitbucket.jenkins.internal.listener;
+
+import com.atlassian.bitbucket.jenkins.internal.client.BitbucketClientFactoryProvider;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketPluginConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketServerConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.model.BitbucketBuildStatus;
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCM;
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMException;
+import com.atlassian.bitbucket.jenkins.internal.utils.CredentialUtils;
+import hudson.Extension;
+import hudson.model.*;
+import hudson.model.listeners.RunListener;
+
+import javax.inject.Inject;
+
+@Extension
+public class BitbucketBuildListener<R extends Run> extends RunListener<R> {
+
+    @Inject
+    private BitbucketClientFactoryProvider bitbucketClientFactoryProvider;
+    @Inject
+    private BitbucketPluginConfiguration pluginConfiguration;
+
+    @Override
+    public void onCompleted(R r, TaskListener listener) {
+        if (r instanceof FreeStyleBuild) {
+            FreeStyleBuild build = (FreeStyleBuild) r;
+            if (build.getProject().getScm() instanceof BitbucketSCM) {
+                BitbucketSCM scm = (BitbucketSCM) build.getProject().getScm();
+                BitbucketServerConfiguration server = pluginConfiguration.getServerById(scm.getServerId())
+                        .orElseThrow(() -> new BitbucketSCMException("Error here"));
+                bitbucketClientFactoryProvider.getClient(server, CredentialUtils.getCredentials(server.getCredentialsId()))
+                        .getBuildStatusClient(scm.getLatestRevision(build))
```

Comment:
Can the latest version be retrieved from the environment?
```
build.getEnvironment(listener).get(GitSCM.GIT_COMMIT);
```

Reply:
yes, I would prefer we do not rely on environment if we can.
We can get hold of it via
`build.getAction(BuildData.class).getLastBuiltRevision().getSha1String()`
where `build` is an instance of `Run` or `R` as it has so nicely been named in this class

Reply:
Why don't you prefer environment?

Reply:
The problem with `r.getAction(BuildData.class).getLastBuiltRevision().getSha1String()` is that it's not guaranteed to be the current build.

Reply:
...yes it is. on line1103 in the GitSCM (in `determineRevisionToBuild(..)` we have this line
`buildData.saveBuild(revToBuild);`
a few lines down we have
`return revToBuild`
_later_ in the class we have
` environment.put(GIT_COMMIT, revToBuild.revision.getSha1String());`
if we look at the `BuildData` we have
```java
 @Exported
    public @CheckForNull Revision getLastBuiltRevision() {
        return lastBuild==null?null:lastBuild.revision;
    }
```
SO, if the environment has been set, so has the 'lastBuiltRevision` as it is updated **before** the environment is set. So they are equally up to date. (I realise the examples are a bit sparse, and there are a few jumps, but do have a look at what happens in the various methods.

Reply:
What happens if parallel build are run? Are you sure that `BuildData` is unique for a run and not for a build?

Reply:
Is there a reason we prefer `build.getAction(BuildData.class).getLastBuildRevision()` to `gitSCM.getBuildData(build).lastBuild.getRevision()`? Quick glance over the code it looks like much of a muchness but I might be missing something